### PR TITLE
[batch] cleanup + delete pods

### PR DESCRIPTION
**Pull request by danking**
cc: @cseed

I wanted to delete pods when they were done, but I felt we needed to be more
careful in general with how we handled k8s API calls. I introduce a little
wrapper to the k8s api that coerces exceptions into return values, unblocks
network calls, and sets some default kwargs.

I also delete pods if we no longer need them. The only reason to keep a pod
around is if it is running or we haven't durably stored the logs.

## K8s class
1. convert blocking calls to non-blocking ones
2. set default kwargs
3. convert ApiExceptions to return values

Instead of the following (which we forgot to make non-blocking):
```python
try:
    v1.delete_namespaced_persistent_volume_claim(
        self._pvc_name,
        HAIL_POD_NAMESPACE,
        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
except kube.client.rest.ApiException as err:
    if err.status == 404:
        log.info(f'persistent volume claim {self._pvc_name} is already deleted')
        return
    raise
finally:
    await db.jobs.update_record(self.id, pvc_name=None)
    self._pvc_name = None
```
We do:
```python
err = await app['k8s'].delete_pvc(self._pvc_name)
if err
    if err.status != 404:
        raise ValueError('could not delete pvc {self._pvc_name}') from err
    log.info(f'persistent volume claim {self._pvc_name} is already deleted')
await db.jobs.update_record(self.id, pvc_name=None)
self._pvc_name = None
```

### CODE REVIEW

**Comment by danking** [comment]
@jigold to address your points from before:
1. Pairing db update and/or self._pod_name with the k8s call, this might be a good idea but I think is orthogonal to this change. I want to focus on cleaning up k8s use and getting pods deleted, then think a little harder about how we should restructure the code more generally to be more understandable
2. I always return at least the error so that the calling code has to decide what an error means
3. I now always delete the pod after the db update, which I think is right b/c we don't want the refresh loop to recreate the pod.

File: batch/batch/server/k8s.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+class K8s:
+    def __init__(self, blocking_pool, timeout, namespace, k8s_api):
+        self.blocking_pool = blocking_pool
+        self.timeout = timeout
+        self.namespace = namespace
+        self.__delete_pod = self._wrap_k8s_delete(k8s_api.delete_namespaced_pod)
+        self.__delete_pvc = self._wrap_k8s_delete(k8s_api.delete_namespaced_persistent_volume_claim)
+        self.__create_pod = self._wrap_k8s(k8s_api.create_namespaced_pod)
+        self.__create_pvc = self._wrap_k8s(k8s_api.create_namespaced_persistent_volume_claim)
+        self.__read_pod_log = self._wrap_k8s(k8s_api.read_namespaced_pod_log)
```
Comment (jigold) [inline batch/batch/server/k8s.py]: Why are these double `__`?

File: batch/batch/server/server.py [hunk 2]
```diff
@@ -216,56 +214,36 @@ def _has_next_task(self):
...
-            await db.jobs.update_record(self.id,
-                                        pod_name=self._pod_name)
-
-            log.info('created pod name: {} for job {}, task {}'.format(self._pod_name,
-                                                                       self.id,
-                                                                       self._current_task.name))
-        except kube.client.rest.ApiException as err:
+        pod, err = await app['k8s'].create_pod(body=pod_template)
+        if err is not None:
+            traceback.print_tb(err.__traceback__)
```
Comment (jigold) [inline batch/batch/server/server.py]: What is the traceback for? Isn't it sufficient in the log statement below?

File: batch/batch/server/server.py [hunk 3]
```diff
@@ -216,56 +214,36 @@ def _has_next_task(self):
...
-                    _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
-            except kube.client.rest.ApiException as err:
-                if err.status == 404:
-                    pass
-                raise
-            finally:
-                await db.jobs.update_record(self.id, pod_name=None)
-                self._pod_name = None
+            await db.jobs.update_record(self.id, pod_name=None)
+            await app['k8s'].delete_pod(name=self._pod_name,
```
Comment (jigold) [inline batch/batch/server/server.py]: Why do we not check or log the error code here?

File: batch/batch/server/server.py [hunk 4]
```diff
@@ -278,37 +256,32 @@ def _has_next_task(self):
...
         uri = None
         if log is not None:
             uri = await write_gs_log_file(app['blocking_pool'], INSTANCE_ID, self.id, task_name, log)
 
         await db.jobs.update_with_log_ec(self.id, task_name, uri, exit_code,
                                          task_idx=self._task_idx,
-                                         pod_name=self._pod_name,
+                                         pod_name=None,
                                          duration=self.duration)
+        await app['k8s'].delete_pod(self._pod_name)
```
Comment (jigold) [inline batch/batch/server/server.py]: Same question about exception and logging as above

File: batch/batch/server/server.py [hunk 5]
```diff
@@ -480,10 +453,12 @@ def is_successful(self):
     async def mark_unscheduled(self):
         if self._pod_name:
             await db.jobs.update_record(self.id, pod_name=None)
+            await app['k8s'].delete_pod(self._pod_name)
```
Comment (jigold) [inline batch/batch/server/server.py]: Same question

File: batch/batch/server/server.py [hunk 6]
```diff
@@ -500,17 +475,12 @@ def is_successful(self):
...
-                    _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
-            except kube.client.rest.ApiException as exc:
-                log.exception(f'could not get logs for {pod.metadata.name} due to {exc}')
-                if exc.status == 400:
-                    await self.mark_unscheduled()
-                    return
-                raise
+            pod_log, err = await app['k8s'].read_pod_log(pod.metadata.name)
+            if err:
+                traceback.print_tb(err.__traceback__)
```
Comment (jigold) [inline batch/batch/server/server.py]: Same question.

File: batch/batch/server/server.py [hunk 7]
```diff
@@ -1098,7 +1051,7 @@ def __anext__(self):
 def serve(port=5000):
     app.add_routes(routes)
     with concurrent.futures.ThreadPoolExecutor() as pool:
-        app['blocking_pool'] = pool
+        app['k8s'] = K8s(pool, KUBERNETES_TIMEOUT_IN_SECONDS, HAIL_POD_NAMESPACE, v1)
```
Comment (jigold) [inline batch/batch/server/server.py]: I think I used the 'blocking_pool' for writing/ reading to google storage.

File: batch/batch/server/server.py (hunk 2 above)
Comment (danking) [reply batch/batch/server/server.py]: The log statement below doesn't print a traceback, just the error message.

File: batch/batch/server/k8s.py (hunk 1 above)
Comment (danking) [reply batch/batch/server/k8s.py]: fixed

File: batch/batch/server/server.py (hunk 3 above)
Comment (danking) [reply batch/batch/server/server.py]: I went through and made sure we at least logged errors when they occurred. There's a few cases where we should probably think more carefully about what an error means.

**Comment by danking** [comment]
@jigold I added `LogStore` and `GCS` to manage non-blocking interaction with GCS.

File: batch/batch/server/server.py [hunk 8]
```diff
@@ -1095,6 +1070,8 @@ def serve(port=5000):
     app.add_routes(routes)
     with concurrent.futures.ThreadPoolExecutor() as pool:
         app['blocking_pool'] = pool
```
Comment (jigold) [inline batch/batch/server/server.py]: Is blocking pool still used now?

Comment (danking) [reply batch/batch/server/server.py]: Yes, by `DeblockedIterator`.

File: batch/batch/server/server.py [hunk 9]
```diff
@@ -216,104 +215,97 @@ def _has_next_task(self):
...
 
         if self._state == 'Ready':
             if self._pod_name:
-                try:
-                    log = v1.read_namespaced_pod_log(
-                        self._pod_name,
-                        HAIL_POD_NAMESPACE,
-                        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+                log, err = await app['k8s'].read_pod_log(self._pod_name)
+                if err is not None:
```
[the remark this answers is not in the document]
Comment (danking) [reply batch/batch/server/server.py]: fixed and added better logging when there is an error

File: batch/batch/server/server.py [hunk 10]
```diff
@@ -216,104 +215,97 @@ def _has_next_task(self):
...
-            if err.status == 404:
-                log.info(f'persistent volume claim {self._pvc_name} is already deleted')
-                return
-            raise
-        finally:
-            await db.jobs.update_record(self.id, pvc_name=None)
-            self._pvc_name = None
+        err = await app['k8s'].delete_pvc(self._pvc_name)
+        if err is not None:
+            raise ValueError('could not delete {self._pvc_name}') from err
```
[the remark this answers is not in the document]
Comment (danking) [reply batch/batch/server/server.py]: the `from err` indicates that this value error is caused by `err`. `err`'s stack trace will be included wherever the `ValueError` is handled.

Good point though, I changed this to have the same behavior as `_delete_k8s_resources`: all errors are ignored but logged.

File: batch/batch/server/server.py [hunk 11]
```diff
@@ -216,104 +215,97 @@ def _has_next_task(self):
...
+                log.info(f'ignoring pod deletion failure for job {self.id} due to {err}')
+            self._pod_name = None
 
     async def _read_logs(self):
         async def _read_log(jt):
             log_uri = await db.jobs.get_log_uri(self.id, jt.name)
-            return jt.name, await read_gs_log_file(app['blocking_pool'], log_uri)
+            if log_uri is None:
+                return None
+            log, err = app['log_store'].read_gs_log_file(log_uri)
```
[the remark this answers is not in the document]
Comment (danking) [reply batch/batch/server/server.py]: fixed

File: batch/batch/server/log_store.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        uri = uri.lstrip('gs://').split('/')
+        bucket_name = uri[0]
+        path = '/'.join(uri[1:])
+        return self.gcs.download_gs_file_as_string(bucket_name, path)
+
+    async def delete_gs_log_file(self, job_id, task_name):
+        path = self._gs_log_path(job_id, task_name)
+        err = self.gcs.delete_gs_file(self.batch_bucket_name, path)
+        if isinstance(err, google.api_core.exceptions.NotFound):
+            self.log.info(f'ignoring: cannot delete nonextant log file: {err}')
```
Comment (jigold) [inline batch/batch/server/log_store.py]: Can you please fix this typo?

File: batch/batch/server/server.py [hunk 13]
```diff
@@ -216,104 +215,104 @@ def _has_next_task(self):
...
-            except kube.client.rest.ApiException as err:
-                if err.status == 404:
-                    pass
-                raise
-            finally:
-                await db.jobs.update_record(self.id, pod_name=None)
-                self._pod_name = None
+            await db.jobs.update_record(self.id, pod_name=None)
+            err = await app['k8s'].delete_pod(name=self._pod_name,
+                                              namespace=HAIL_POD_NAMESPACE)
```
Comment (jigold) [inline batch/batch/server/server.py]: Is this a different namespace than the default?

File: batch/batch/server/log_store.py (hunk 12 above)
Comment (danking) [reply batch/batch/server/log_store.py]: Unless I'm missing something, I assume you're referring to nonextant? https://www.merriam-webster.com/dictionary/nonextant But I can reword

Comment (jigold) [reply batch/batch/server/log_store.py]: I thought it should be nonexistent.

File: batch/batch/server/google_storage.py [hunk 14]
```diff
@@ -44,7 +44,7 @@ def _wrap_returning_network_call(self, fun):
         return wrapped
 
     def _wrap_nonreturning_network_call(self, fun):
-        fun = self._wrap_nonreturning_network_call(fun)
+        fun = self._wrap_returning_network_call(fun)
```
Comment (jigold) [inline batch/batch/server/google_storage.py]: The function names don't match now...

Comment (danking) [reply batch/batch/server/google_storage.py]: They're not supposed to, this was an infinite loop before. `_wrap_returning_network_call` does the normal wrapping stuff. `_wrap_nonreturning_network_call` just drops the returned value.

Comment (jigold) [reply batch/batch/server/google_storage.py]: ok.

**Pull request merged**